pkg/admissioncontroller/server/handlers/webhooks/validate_namespace_deletion_test.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhooks_test
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/gardener/gardener/pkg/admissioncontroller/server/handlers/webhooks"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	coreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/kubernetes/fake"
+	mockcache "github.com/gardener/gardener/pkg/mock/controller-runtime/cache"
+	mockclient "github.com/gardener/gardener/pkg/mock/controller-runtime/client"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("ValidateNamespaceDeletion", func() {
+	var (
+		ctx       = context.Background()
+		namespace = &corev1.Namespace{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Namespace",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "foo",
+			},
+		}
+	)
+
+	DescribeTable("Namespace deletion admission",
+		func(namespace *corev1.Namespace, cl client.Client, expectedAllowed bool, expectedMsg string) {
+			ctrl := gomock.NewController(GinkgoT())
```

Comment:
I find it a bit confusing, that you create a lot of mocks here, for which calls are never verified. So I assume you only mock the clients in order to have some preexisting objects.

If that's correct and you don't need to check, that explicit calls are made, I would replace the `mockClient` by [fake clients](https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg/client/fake) and the mock cache and `syncedInformer` by [fake informers](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/informertest/fake_cache.go).
We shouldn't need to reimplement the stuff already available upstream, if it's sufficient for our use cases.

Another option without mocking or faking anything would be to use a local control plane (envtest).

WDYT?

Reply:
I was switching to the fakeClient when I noticed a mock is more appropriate here even if the primary goal is not to verify the calls. With mocks it's easily possible to specify or configure the behavior, especially for errors apart from `NotFound`. But as you said, not verifying the calls was awkward, so I'm now calling `ctrl.Finish()` at the end of every test case.